src/basicfuns.jl

```diff
@@ -308,10 +308,25 @@ $(SIGNATURES)
 Return `log(exp(x) + exp(y))`, avoiding intermediate overflow/undeflow, and handling
 non-finite values.
 """
-function logaddexp(x::Real, y::Real)
-    # ensure Δ = 0 if x = y = ± Inf
-    Δ = x == y ? zero(x - y) : abs(x - y)
-    max(x, y) + log1pexp(-Δ)
+logaddexp(x::Real, y::Real) = logaddexp(float.(promote(x, y))...)
+
+function logaddexp(x::T, y::T) where T <: AbstractFloat
+    if x < y
+        diff = x - y
+        maxval = y
+    elseif y <= x
+        diff = y - x
+        maxval = x
+    else
+        diff = zero(T)
+        maxval = T(NaN)
+    end
```

Comment:
This seems quite complicated. Is it helpful for improving performance?

Reply:
Yes, the point is to have a single branching instead of two when successively calling `max(x, y)` and `abs(x - y)`. Here is the benchmark when there is 2 branching (using `maxval = max(x, y)` and `diff = abs(x - y)`):
```
BenchmarkTools.Trial: 4680 samples with 1 evaluation.
 Range (min … max):  1.053 ms …  1.347 ms  ┊ GC (min … max): 0.00% … 20.78%
 Time  (median):     1.062 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.067 ms ± 28.219 μs  ┊ GC (mean ± σ):  0.34% ±  2.15%
```
and when there is only 1 branching:
```
BenchmarkTools.Trial: 8614 samples with 1 evaluation.
 Range (min … max):  567.500 μs … 827.459 μs  ┊ GC (min … max): 0.00% … 30.88% 
 Time  (median):     571.000 μs               ┊ GC (median):    0.00%            
 Time  (mean ± σ):   579.051 μs ±  33.001 μs  ┊ GC (mean ± σ):  0.70% ±  3.80%     
```